Update existential subquery CIP (#493)
opencypher/openCypher: cip/1.accepted/CIP2015-05-13-EXISTS.adoc

:toc-placement: macro
:source-highlighter: codemirror

*Authors:* Andrés Taylor <andres@neotechnology.com>, Hannes Voigt <hannes.voigt@neo4j.com>

[abstract]
.Abstract
--

> This CIP introduces existential subquery and cleans up previously existing language constructs for testing of existence.

Reviewer: ```suggestion
This CIP introduces existential subqueries and cleans up previously existing language constructs for existence testing.
```

--

toc::[]

== Background

> openCypher has two language constructs that allow testing of existence. Namely

Reviewer: ```suggestion
openCypher has two language constructs that allow testing of existence. 
Namely
```

just for the "one sentence per line" convention

> * property existence predicate: `exists( n.prop )`, which is grammarically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=FunctionInvocation[<FunctionInvocation>]

Reviewer: ```suggestion
* property existence predicate: `exists( n.prop )`, which is grammatically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=FunctionInvocation[<FunctionInvocation>]
```

> * pattern predicates for the existence of subgraphs, which is grammarically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>]

Reviewer: ```suggestion
* pattern predicates for the existence of subgraphs, which is grammatically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>]
```

> Both are allowed as https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Atom[<Atom>]s of expressions.

Reviewer: these links are a treat!

> Note that a <RelationshipsPattern> appearing in an expression is evaluated to a list containing the matches of the pattern.

Reviewer: ```suggestion
Note that a <RelationshipsPattern> appearing in an expression is evaluated to a list of paths containing the matches of the pattern.
```

In a predicate context this list gets coerced to a boolean (empty list -> [underline]#*_False_*#, non-empty list -> [underline]#*_True_*#), letting this <RelationshipsPattern> behave like a predicate although it actually is a list expression.
As a result, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>s] appear to behave differently depending on the context in which they appear.

> Further, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[<PatternComprehension>] offer a more powerful and syntaxtically better separated means to   get lists of matches for a pattern within expressions.

Reviewer: ```suggestion
Further, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[<PatternComprehension>] offer a more powerful and syntactically better separated means to get lists of matches for a pattern within expressions.
```

The understanding of subqueries in Cypher has evolved and Cypher gained a proper https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>] (`IS [ NOT ] NULL`), cf. https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL].
Both developments and the issues around <RelationshipsPattern>s suggest the need for a refinement of the language constructs that allow testing of existence.

== Proposal

> CIP proposes:

Reviewer: ```suggestion
This CIP proposes:
```

> * Restricting https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] to predicate contexts and boolean evaluation

Reviewer: They are of course still valid to use in clauses such as MATCH and MERGE. Perhaps we ought to specify explicitly here that we mean within the remits of expressions?
```suggestion
* Restricting https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] to predicate contexts and boolean evaluation, when used in expression contexts
```

Reviewer: or perhaps this is not applicable because `RelationshipsPattern` is not directly referenced in these clauses, but is only referenced by way of `Pattern`, and so there's no formal need to mention this here?

Author: That's right. RelationshipsPattern is only used in Atom (expressions) and PatternComprehension. So your suggestion still makes sense. :-)

* Adding two forms of existential subqueries denoted with `EXISTS { ... }`:
** Full existential subquery based on a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>]
** Simple existential subquery based on a single https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Pattern[<Pattern>] and an optional https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Where[<Where>] clause

> * Removing of `exists( n.prop )` in favor of the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>], which is a restatement of what has been accepted with https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL]

Reviewer: ```suggestion
* Removing `exists( n.prop )` in favor of the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>], which is a restatement of what has been accepted with https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL]
```

=== Syntax

==== Grammar

[source,bnf]
----
<Atom> ::=
    ...
  | <RelationshipsPattern>
  | <ExistentialSubquery>

<ExistentialSubquery> ::=
    <SimpleExistentialSubquery>
  | <FullExistentialSubquery>

> <SimpleExistentialSubquery> ::=
>   "EXISTS" "{" <SimpleMatch> "}"

Reviewer: maybe it's a dialectal difference, but the EBNF I've learned uses comma separation between parts
```suggestion
<SimpleExistentialSubquery> ::=
  "EXISTS", "{", <SimpleMatch>, "}"
```

Author: done.

** <Where>, used in
*** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Match[<Match>]
*** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=With[<With>]
*** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=YieldItems[<YieldItems>]
*** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=FilterExpression[<FilterExpression>], used in
**** `ALL`, `ANY`, `NONE`, `SINGLE`
**** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ListComprehension[<ListComprehension>]
*** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[<PatternComprehension>]

> ** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=CaseAlternatice[<CaseAlternative>]

Reviewer: ```suggestion
** https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=CaseAlternative[<CaseAlternative>]
```

> * The https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>] contained in a <FullExistentialSubquery> shall not contain any https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=UpdatingClause[<UpdatingClause>] nor procedure or function calls with any side effects

Reviewer: question on the definition of "syntax rule": I understand that updating clauses are syntactically identifiable, but procs and functions _with side effects_ are not. Do these still belong under this heading?

Author: That is debatable either way. I do not have strong opinions on it. My assumption is that the implementation would statically know if something may cause side effects, which suggests a slight refined wording though:

```suggestion
* The https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>] contained in a <FullExistentialSubquery> shall not contain any https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=UpdatingClause[<UpdatingClause>] nor procedure or function calls that are not known to be free of side effects.
```

=== Semantics

https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>], <SimpleExistentialSubquery>, and <FullExistentialSubquery> accomplish the same task: checking whether the result of a subquery is not empty.

All three forms,

* Are boolean expressions, i.e. return a single boolean value
* Never return `null`
* Are side effect-free, i.e. <FullExistentialSubquery> shall not contain any <UpdatingClause> or other sources of side effects

> * Can contain variables and parameters from the outer queries

Reviewer: nice listing!

> https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall not to introduce new variables.

Reviewer: ```suggestion
https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall not be allowed to introduce new variables.
```

Author: The sentences is off, I agree, but I fix it slightly differently.

```suggestion
https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall not introduce new variables.
```

Both forms of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ExistentialSubquery[<ExistentialSubquery>] (<SimpleExistentialSubquery> and <FullExistentialSubquery>) are allowed to introduce new variables.
These variables necessarily shall have a name different from the names of all variables available from the outer queries.

> Any variables introduced in an <ExistentialSubquery> are not available on the outside.

Reviewer: ```suggestion
Any variables introduced in an <ExistentialSubquery> are not available outside the subquery context.
```

https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] and <SimpleExistentialSubquery> are syntactically simpler and semantically less powerful forms of <FullExistentialSubquery>.

> The semantics of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] and <SimpleExistentialSubquery> can be defined as syntax transformations to <FullExistentialSubquery>.

Reviewer: yes!

==== <RelationshipsPattern>

A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] _RP_ is effectively replaced by the expression

> `EXISTS { MATCH _RP_ RETURN 1 }`

Reviewer: RETURN 1 is arbitrary, but I don't know of a less arbitrary alternative. I considered commenting we should write `RETURN true` here instead to stay with booleans, but perhaps that could be misinterpreted as being an important return value. No return value is important; the presence of any row indicates existence. It would be clean to have some kind of "return a row with no fields" construct.

`RETURN *` also does not work since RPs cannot introduce variables so it could fail.

Author: Good observation and thanks for running through a number of alternative. I hadn't really do that.

Also: It just specifies what is effectively suppose to happen. How an implementation precisely does it is up to the implementation. Hence, it is also not super important which of the arbitrary values the specification uses as long as it is not wrong or terribly misleading.

Reviewer: Agreed. But the ideal thing would be for the spec to mention nothing that is irrelevant; unfortunately we are unable to achieve that right here -- your choice is the best though.

==== <SimpleExistentialSubquery>

A <SimpleExistentialSubquery> containing a <SimpleMatch> _SM_ is effectively replaced by the expression

`EXISTS { MATCH _SM_ RETURN 1 }`

==== <FullExistentialSubquery>

A <FullExistentialSubquery> _FES_ is effectively evaluated as follows:

* Let _OUTER_VARIABLES_ be the current working record for which the expression containing _FES_ is evaluated.
* Let _NESTED_QUERY_ be the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>] immediately contained in _FES_.
* Let _RESULT_TABLE_ be the table resulting from evaluating _NESTED_QUERY_ on a driving table comprising _OUTER_VARIABLES_.
* Case:
** If _RESULT_TABLE_ is an empty table (cardinality is zero), then the result of _FES_ is [underline]#*_False_*#.

> ** Otherwise, the result of _FES_ is [underline]#*_True_*#.

Reviewer: This explanation is excellent! Just as a detail, perhaps we ought to call out explicitly that all fields of the resulting table are ignored?

Author: That GQL spec lingo. You would not say anything else. _RESULT_TABLE_ is a spec variable which implicitly "dies" with the end of the rule block. However, we can add an informative note w.r.t you point.

```suggestion
** Otherwise, the result of _FES_ is [underline]#*_True_*#.

Note that all fields in _RESULT_TABLE_ are ignored and only the number of rows in _RESULT_TABLE_ is relevant for the result of _FES_.
```

=== Examples

==== Property existence test

_Example 1:_

Return all nodes that have a property named `slogan`.
[source, cypher]
----
MATCH (actor)
WHERE actor.slogan IS NOT NULL
RETURN actor
----

==== Pattern predicates in boolean expression context

_Example 2A:_

Find all actors who won an award.

[source, cypher]
----
MATCH (actor:Actor) WHERE (actor)-[:WON]->(:Award)
RETURN actor
----

_Example 2B:_

Find all actors with their major accolade.

[source, cypher]
----
MATCH (actor:Actor)
RETURN actor,
  CASE actor
    WHEN (actor)-[:WON]->(:Oscar) THEN 'Oscar winner'
    WHEN (actor)-[:WON]->(:GoldenGlobe) THEN 'Golden Globe winner'
    ELSE 'None'
  END AS accolade
----

_Example 2C:_

Find all movies that have at least one award-winning actor in their cast.

[source, cypher]
----
MATCH (movie:Movie)<-[:ACTED_IN]-(actor:Actor)
WITH movie, collect(actor) AS cast
WHERE ANY(actor IN cast WHERE (actor)-[:WON]->(:Award))
RETURN movie
----

> 

Reviewer: great examples!

:toc-placement: macro
:source-highlighter: codemirror

*Authors:* Andrés Taylor <andres@neotechnology.com>, Hannes Voigt <hannes.voigt@neo4j.com>

[abstract]
.Abstract
--
This CIP introduces existential subqueries and cleans up previously existing language constructs for existence testing.
--

toc::[]

== Background

openCypher has two language constructs that allow testing of existence. 
Namely

* property existence predicate: `exists( n.prop )`, which is grammatically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=FunctionInvocation[<FunctionInvocation>]
* pattern predicates for the existence of subgraphs, which is grammatically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>]

Both are allowed as https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Atom[<Atom>]s of expressions.

> Note that a <RelationshipsPattern> appearing in an expression is evaluated to a list of paths containing the matches of the pattern.

Reviewer 2: Is `<RelationshipsPattern>` supposed to be a link like below? (Twice in this paragraph)

In a predicate context this list gets coerced to a boolean (empty list -> [underline]#*_False_*#, non-empty list -> [underline]#*_True_*#), letting this <RelationshipsPattern> behave like a predicate although it actually is a list expression.
As a result, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>s] appear to behave differently depending on the context in which they appear.
Further, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[<PatternComprehension>] offer a more powerful and syntactically better separated means to get lists of matches for a pattern within expressions.

The understanding of subqueries in Cypher has evolved and Cypher gained a proper https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>] (`IS [ NOT ] NULL`), cf. https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL].

> Both developments and the issues around <RelationshipsPattern>s suggest the need for a refinement of the language constructs that allow testing of existence.

Reviewer 2: Is `<RelationshipsPattern>` supposed to be a link?

Author: I have added the links.

* Removing `exists( n.prop )` in favor of the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>], which is a restatement of what has been accepted with https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL]

=== Syntax

==== Grammar

[source,bnf]
----
<Atom> ::=
    ...
  | <RelationshipsPattern>
  | <ExistentialSubquery>

<ExistentialSubquery> ::=
    <SimpleExistentialSubquery>
  | <FullExistentialSubquery>

<SimpleExistentialSubquery> ::=
  "EXISTS", "{", <SimpleMatch>, "}"

<FullExistentialSubquery> ::=
  "EXISTS", "{", <RegularQuery>, "}"

<SimpleMatch> ::=
  <Pattern>, [ <Where> ]
----

Note that the openCypher grammar does not list <SimpleExistentialSubquery>, <SimpleMatch>, and <FullExistentialSubquery> as separate productions but represents them inline of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ExistentialSubquery[<ExistentialSubquery>].

==== Syntax Rules

> * A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be contained in sites whose expected type is exactly boolean. Specifically, a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be simply contained in

Reviewer 2: This is in direct conflict with a previous CLG decision to allow `EXISTS {  }` as a general expression: https://trello.com/c/yUwor3uw/532-allow-exists-as-an-expression

Reviewer: This line doesn't say anything about `EXISTS { }`, does it? It talks about the "pattern expression" form, ie `WITH ()-->() AS boolean`-style things.

Later on it says

>All three forms ... are boolean expressions

So I don't really see a conflict?

Reviewer: Specifically, what I read from this line is that we cannot do `size( ()-->() )`, because `size()` does not expect a boolean. Hence the other discussion we had in CLG last meeting.

Reviewer 2: Right, I misread that. In that case I retract my comment about the conflict.

On a different note, I think the following wording is confusing.
> A RelationshipsPattern shall only be contained in sites whose expected type is exactly boolean. Specifically, a RelationshipsPattern shall only be simply contained in Where, ...

There are more places than listed here where the expected type is exactly boolean. For example `RETURN NOT ()--()`. We deprecated these use cases and I expect they would be forbidden in this proposal, but the wording is unclear.

Reviewer 2: I don't think we need to list every grammar rule that has exactly the expected type boolean. What do you think about rephrasing like this or similar?

```suggestion
* A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be contained in sites whose expected type is exactly boolean. To give the most frequent examples, a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] can be simply contained in
```

* Let _NESTED_QUERY_ be the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>] immediately contained in _FES_.
* Let _RESULT_TABLE_ be the table resulting from evaluating _NESTED_QUERY_ on a driving table comprising _OUTER_VARIABLES_.
* Case:
** If _RESULT_TABLE_ is an empty table (cardinality is zero), then the result of _FES_ is [underline]#*_False_*#.
** Otherwise, the result of _FES_ is [underline]#*_True_*#.

Note that all fields in _RESULT_TABLE_ are ignored and only the number of rows in _RESULT_TABLE_ is relevant for the result of _FES_.

=== Examples

==== Property existence test

_Example 1:_

Return all nodes that have a property named `slogan`.
[source, cypher]
----
MATCH (actor)
WHERE actor.slogan IS NOT NULL
RETURN actor
----

==== Pattern predicates in boolean expression context

_Example 2A:_

Find all actors who won an award.

[source, cypher]
----
MATCH (actor:Actor) WHERE (actor)-[:WON]->(:Award)
RETURN actor
----

_Example 2B:_

Find all actors with their major accolade.

[source, cypher]
----
MATCH (actor:Actor)
RETURN actor,

>   CASE actor

Reviewer 2: ```suggestion
  CASE
```

When using the [Simple case form](https://neo4j.com/docs/cypher-manual/4.2/syntax/expressions/#syntax-simple-case) the expected type after `WHEN` is not exactly boolean.

Author: You are absolutely right. I meant the full case form but managed to get the syntax wrong.

This proposal also allows for powerful subqueries, for example using aggregation inside the `EXISTS {}` query.

> Find all teams that have at least two members who have worked on successful projects.

Reviewer: how come you removed the leading dot?

Author: Do not remember exactly, but think it was not used consistently in similar spots.

Reviewer: OK. It is used to render above the listing (the Cypher code in this case) in a special way (not just a paragraph). I kinda like it but no objection here.

RETURN team
----

However, pattern predicates have a readability advantage in narrow cases.

> Hence, this proposal retains them while removing their confusing meaning outside boolean expression context.

Reviewer: 👍
